File: backend/src/main/java/com/neobit/crm/controller/IntegrationController.java
package com.neobit.crm.controller;

import com.neobit.crm.dto.request.*;
import com.neobit.crm.dto.response.*;
import com.neobit.crm.security.TenantContext;
import com.neobit.crm.security.UserPrincipal;
import com.neobit.crm.service.integration.*;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDate;
import java.util.List;
import java.util.UUID;

/**
 * Integration Controller
 * 
 * Handles third-party integrations:
 * - Gmail (email sync)
 * - Google Calendar (event sync)
 * - Telegram Bot (messaging)
 * - ClickUp (task management)
 * - ZegoCloud (voice/video)
 */
@RestController
@RequestMapping("/api/integrations")
@RequiredArgsConstructor
@Slf4j
@Tag(name = "Integrations", description = "Third-party integration endpoints")
public class IntegrationController {

    private final GmailService gmailService;
    private final CalendarService calendarService;
    private final TelegramService telegramService;
    private final ClickUpService clickUpService;
    private final ZegoService zegoService;
    private final IntegrationConfigService configService;

    // ============================================================
    // GENERAL INTEGRATION MANAGEMENT
    // ============================================================

    /**
     * List all configured integrations
     * 
     * GET /api/integrations
     * 
     * Response:
     * {
     *   "integrations": [
     *     { "type": "GMAIL", "status": "CONNECTED", "connectedEmail": "..." },
     *     { "type": "TELEGRAM", "status": "CONNECTED", "botUsername": "@bot" }
     *   ]
     * }
     */
    @GetMapping
    @PreAuthorize("hasAnyRole('VENDOR_ADMIN', 'PLATFORM_ADMIN')")
    @Operation(summary = "List configured integrations")
    public ResponseEntity<IntegrationsListResponse> listIntegrations() {
        
        UUID tenantId = TenantContext.getTenantId();
        
        IntegrationsListResponse response = configService.listIntegrations(tenantId);
        
        return ResponseEntity.ok(response);
    }

    /**
     * Get integration status
     * 
     * GET /api/integrations/{type}
     */
    @GetMapping("/{type}")
    @PreAuthorize("hasAnyRole('VENDOR_ADMIN', 'PLATFORM_ADMIN')")
    @Operation(summary = "Get integration status")
    public ResponseEntity<IntegrationStatusResponse> getIntegrationStatus(
            @PathVariable String type) {
        
        UUID tenantId = TenantContext.getTenantId();
        
        IntegrationStatusResponse status = configService.getIntegrationStatus(
            tenantId, type.toUpperCase()
        );
        
        return ResponseEntity.ok(status);
    }

    /**
     * Connect integration (initiate OAuth)
     * 
     * POST /api/integrations/{type}/connect
     */
    @PostMapping("/{type}/connect")
    @PreAuthorize("hasAnyRole('VENDOR_ADMIN', 'PLATFORM_ADMIN')")
    @Operation(summary = "Initiate integration connection")
    public ResponseEntity<OAuthUrlResponse> connectIntegration(
            @PathVariable String type,
            @RequestBody(required = false) ConnectIntegrationRequest request) {
        
        log.info("Initiating connection for integration: {}", type);
        
        UUID tenantId = TenantContext.getTenantId();
        String redirectUri = request != null ? request.getRedirectUri() : null;
        
        String authUrl = configService.getOAuthUrl(tenantId, type.toUpperCase(), redirectUri);
        
        return ResponseEntity.ok(new OAuthUrlResponse(authUrl));
    }

    /**
     * OAuth callback for integration
     * 
     * POST /api/integrations/{type}/callback
     */
    @PostMapping("/{type}/callback")
    @PreAuthorize("hasAnyRole('VENDOR_ADMIN', 'PLATFORM_ADMIN')")
    @Operation(summary = "Handle OAuth callback")
    public ResponseEntity<IntegrationStatusResponse> handleCallback(
            @PathVariable String type,
            @Valid @RequestBody OAuthCallbackRequest request) {
        
        log.info("Processing OAuth callback for: {}", type);
        
        UUID tenantId = TenantContext.getTenantId();
        
        IntegrationStatusResponse status = configService.handleOAuthCallback(
            tenantId, type.toUpperCase(), request.getCode(), request.getState()
        );
        
        return ResponseEntity.ok(status);
    }

    /**
     * Disconnect integration
     * 
     * DELETE /api/integrations/{type}/disconnect
     */
    @DeleteMapping("/{type}/disconnect")
    @PreAuthorize("hasAnyRole('VENDOR_ADMIN', 'PLATFORM_ADMIN')")
    @Operation(summary = "Disconnect integration")
    public ResponseEntity<IntegrationStatusResponse> disconnectIntegration(
            @PathVariable String type) {
        
        log.warn("Disconnecting integration: {}", type);
        
        UUID tenantId = TenantContext.getTenantId();
        
        IntegrationStatusResponse status = configService.disconnect(
            tenantId, type.toUpperCase()
        );
        
        return ResponseEntity.ok(status);
    }

    /**
     * Update integration settings
     * 
     * PUT /api/integrations/{type}/settings
     */
    @PutMapping("/{type}/settings")
    @PreAuthorize("hasAnyRole('VENDOR_ADMIN', 'PLATFORM_ADMIN')")
    @Operation(summary = "Update integration settings")
    public ResponseEntity<IntegrationStatusResponse> updateSettings(
            @PathVariable String type,
            @Valid @RequestBody UpdateIntegrationSettingsRequest request) {
        
        UUID tenantId = TenantContext.getTenantId();
        
        IntegrationStatusResponse status = configService.updateSettings(
            tenantId, type.toUpperCase(), request.getSettings()
        );
        
        return ResponseEntity.ok(status);
    }

    // ============================================================
    // GMAIL INTEGRATION
    // ============================================================

    /**
     * Trigger Gmail sync
     * 
     * POST /api/integrations/gmail/sync
     */
    @PostMapping("/gmail/sync")
    @PreAuthorize("hasAnyRole('VENDOR_ADMIN', 'PLATFORM_ADMIN')")
    @Operation(summary = "Trigger Gmail sync")
    public ResponseEntity<SyncStatusResponse> syncGmail() {
        
        log.info("Starting Gmail sync");
        
        UUID tenantId = TenantContext.getTenantId();
        
        SyncStatusResponse response = gmailService.triggerSync(tenantId);
        
        return ResponseEntity.accepted().body(response);
    }

    /**
     * Get synced emails
     * 
     * GET /api/integrations/gmail/emails
     */
    @GetMapping("/gmail/emails")
    @Operation(summary = "Get synced emails")
    public ResponseEntity<Page<EmailResponse>> getEmails(
            @RequestParam(required = false) UUID customerId,
            @RequestParam(required = false) String label,
            Pageable pageable) {
        
        UUID tenantId = TenantContext.getTenantId();
        
        Page<EmailResponse> emails = gmailService.getEmails(tenantId, customerId, label, pageable);
        
        return ResponseEntity.ok(emails);
    }

    /**
     * Send email via Gmail
     * 
     * POST /api/integrations/gmail/send
     */
    @PostMapping("/gmail/send")
    @Operation(summary = "Send email via Gmail")
    public ResponseEntity<EmailResponse> sendEmail(
            @Valid @RequestBody SendEmailRequest request,
            @AuthenticationPrincipal UserPrincipal principal) {
        
        log.info("Sending email to: {} by user: {}", request.getTo(), principal.getEmail());
        
        UUID tenantId = TenantContext.getTenantId();
        
        EmailResponse response = gmailService.sendEmail(tenantId, principal.getId(), request);
        
        return ResponseEntity.ok(response);
    }

    // ============================================================
    // GOOGLE CALENDAR INTEGRATION
    // ============================================================

    /**
     * Get calendar events
     * 
     * GET /api/integrations/calendar/events
     */
    @GetMapping("/calendar/events")
    @Operation(summary = "Get calendar events")
    public ResponseEntity<List<CalendarEventResponse>> getCalendarEvents(
            @RequestParam(required = false) LocalDate startDate,
            @RequestParam(required = false) LocalDate endDate,
            @RequestParam(required = false) UUID customerId) {
        
        UUID tenantId = TenantContext.getTenantId();
        
        List<CalendarEventResponse> events = calendarService.getEvents(
            tenantId, startDate, endDate, customerId
        );
        
        return ResponseEntity.ok(events);
    }

    /**
     * Create calendar event
     * 
     * POST /api/integrations/calendar/events
     */
    @PostMapping("/calendar/events")
    @Operation(summary = "Create calendar event")
    public ResponseEntity<CalendarEventResponse> createCalendarEvent(
            @Valid @RequestBody CreateCalendarEventRequest request,
            @AuthenticationPrincipal UserPrincipal principal) {
        
        log.info("Creating calendar event: {} by user: {}", 
            request.getTitle(), principal.getEmail());
        
        UUID tenantId = TenantContext.getTenantId();
        
        CalendarEventResponse event = calendarService.createEvent(
            tenantId, principal.getId(), request
        );
        
        return ResponseEntity.ok(event);
    }

    /**
     * Update calendar event
     * 
     * PUT /api/integrations/calendar/events/{eventId}
     */
    @PutMapping("/calendar/events/{eventId}")
    @Operation(summary = "Update calendar event")
    public ResponseEntity<CalendarEventResponse> updateCalendarEvent(
            @PathVariable String eventId,
            @Valid @RequestBody UpdateCalendarEventRequest request) {
        
        UUID tenantId = TenantContext.getTenantId();
        
        CalendarEventResponse event = calendarService.updateEvent(tenantId, eventId, request);
        
        return ResponseEntity.ok(event);
    }

    /**
     * Delete calendar event
     * 
     * DELETE /api/integrations/calendar/events/{eventId}
     */
    @DeleteMapping("/calendar/events/{eventId}")
    @Operation(summary = "Delete calendar event")
    public ResponseEntity<Void> deleteCalendarEvent(@PathVariable String eventId) {
        
        UUID tenantId = TenantContext.getTenantId();
        
        calendarService.deleteEvent(tenantId, eventId);
        
        return ResponseEntity.noContent().build();
    }

    // ============================================================
    // TELEGRAM INTEGRATION
    // ============================================================

    /**
     * Configure Telegram bot
     * 
     * POST /api/integrations/telegram/configure
     */
    @PostMapping("/telegram/configure")
    @PreAuthorize("hasAnyRole('VENDOR_ADMIN', 'PLATFORM_ADMIN')")
    @Operation(summary = "Configure Telegram bot")
    public ResponseEntity<TelegramConfigResponse> configureTelegram(
            @Valid @RequestBody ConfigureTelegramRequest request) {
        
        log.info("Configuring Telegram bot");
        
        UUID tenantId = TenantContext.getTenantId();
        
        TelegramConfigResponse response = telegramService.configure(tenantId, request);
        
        return ResponseEntity.ok(response);
    }

    /**
     * Link Telegram chat to customer
     * 
     * POST /api/integrations/telegram/link
     */
    @PostMapping("/telegram/link")
    @Operation(summary = "Link Telegram chat to customer")
    public ResponseEntity<TelegramLinkResponse> linkTelegramChat(
            @Valid @RequestBody LinkTelegramChatRequest request) {
        
        UUID tenantId = TenantContext.getTenantId();
        
        TelegramLinkResponse response = telegramService.linkChatToCustomer(
            tenantId, request.getChatId(), request.getCustomerId()
        );
        
        return ResponseEntity.ok(response);
    }

    /**
     * Send Telegram message
     * 
     * POST /api/integrations/telegram/send
     */
    @PostMapping("/telegram/send")
    @Operation(summary = "Send Telegram message")
    public ResponseEntity<TelegramMessageResponse> sendTelegramMessage(
            @Valid @RequestBody SendTelegramMessageRequest request,
            @AuthenticationPrincipal UserPrincipal principal) {
        
        UUID tenantId = TenantContext.getTenantId();
        
        TelegramMessageResponse response = telegramService.sendMessage(
            tenantId, principal.getId(), request
        );
        
        return ResponseEntity.ok(response);
    }

    // ============================================================
    // CLICKUP INTEGRATION
    // ============================================================

    /**
     * Get ClickUp workspaces
     * 
     * GET /api/integrations/clickup/workspaces
     */
    @GetMapping("/clickup/workspaces")
    @Operation(summary = "Get ClickUp workspaces")
    public ResponseEntity<List<ClickUpWorkspaceResponse>> getClickUpWorkspaces() {
        
        UUID tenantId = TenantContext.getTenantId();
        
        List<ClickUpWorkspaceResponse> workspaces = clickUpService.getWorkspaces(tenantId);
        
        return ResponseEntity.ok(workspaces);
    }

    /**
     * Create ClickUp task
     * 
     * POST /api/integrations/clickup/tasks
     */
    @PostMapping("/clickup/tasks")
    @Operation(summary = "Create ClickUp task")
    public ResponseEntity<ClickUpTaskResponse> createClickUpTask(
            @Valid @RequestBody CreateClickUpTaskRequest request,
            @AuthenticationPrincipal UserPrincipal principal) {
        
        log.info("Creating ClickUp task: {} by user: {}", 
            request.getName(), principal.getEmail());
        
        UUID tenantId = TenantContext.getTenantId();
        
        ClickUpTaskResponse task = clickUpService.createTask(tenantId, request);
        
        return ResponseEntity.ok(task);
    }

    /**
     * Update ClickUp task
     * 
     * PUT /api/integrations/clickup/tasks/{taskId}
     */
    @PutMapping("/clickup/tasks/{taskId}")
    @Operation(summary = "Update ClickUp task")
    public ResponseEntity<ClickUpTaskResponse> updateClickUpTask(
            @PathVariable String taskId,
            @Valid @RequestBody UpdateClickUpTaskRequest request) {
        
        UUID tenantId = TenantContext.getTenantId();
        
        ClickUpTaskResponse task = clickUpService.updateTask(tenantId, taskId, request);
        
        return ResponseEntity.ok(task);
    }

    /**
     * Get tasks linked to customer
     * 
     * GET /api/integrations/clickup/tasks?customerId=...
     */
    @GetMapping("/clickup/tasks")
    @Operation(summary = "Get ClickUp tasks")
    public ResponseEntity<List<ClickUpTaskResponse>> getClickUpTasks(
            @RequestParam(required = false) UUID customerId,
            @RequestParam(required = false) String status) {
        
        UUID tenantId = TenantContext.getTenantId();
        
        List<ClickUpTaskResponse> tasks = clickUpService.getTasks(tenantId, customerId, status);
        
        return ResponseEntity.ok(tasks);
    }

    // ============================================================
    // ZEGOCLOUD INTEGRATION
    // ============================================================

    /**
     * Generate ZegoCloud token
     * 
     * POST /api/integrations/zego/token
     */
    @PostMapping("/zego/token")
    @Operation(summary = "Generate ZegoCloud token")
    public ResponseEntity<ZegoTokenResponse> generateZegoToken(
            @Valid @RequestBody GenerateZegoTokenRequest request,
            @AuthenticationPrincipal UserPrincipal principal) {
        
        ZegoTokenResponse response = zegoService.generateToken(
            request.getUserId() != null ? request.getUserId() : principal.getId().toString(),
            request.getUserName() != null ? request.getUserName() : principal.getName(),
            request.getRoomId()
        );
        
        return ResponseEntity.ok(response);
    }

    /**
     * Get ZegoCloud configuration
     * 
     * GET /api/integrations/zego/config
     */
    @GetMapping("/zego/config")
    @Operation(summary = "Get ZegoCloud configuration")
    public ResponseEntity<ZegoConfigResponse> getZegoConfig() {
        
        ZegoConfigResponse config = zegoService.getPublicConfig();
        
        return ResponseEntity.ok(config);
    }

    /**
     * Get call history
     * 
     * GET /api/integrations/zego/calls
     */
    @GetMapping("/zego/calls")
    @Operation(summary = "Get call history")
    public ResponseEntity<Page<CallSessionResponse>> getCallHistory(
            @RequestParam(required = false) UUID customerId,
            @RequestParam(required = false) LocalDate startDate,
            @RequestParam(required = false) LocalDate endDate,
            Pageable pageable) {
        
        UUID tenantId = TenantContext.getTenantId();
        
        Page<CallSessionResponse> calls = zegoService.getCallHistory(
            tenantId, customerId, startDate, endDate, pageable
        );
        
        return ResponseEntity.ok(calls);
    }
}

// ============================================================
// Request/Response DTOs (would be in separate files)
// ============================================================

/*
// IntegrationsListResponse.java
@Data
@Builder
public class IntegrationsListResponse {
    private List<IntegrationStatusResponse> integrations;
}

// IntegrationStatusResponse.java
@Data
@Builder
public class IntegrationStatusResponse {
    private String type;
    private String status; // CONNECTED, DISCONNECTED, ERROR
    private String connectedAccount;
    private Map<String, Object> settings;
    private LocalDateTime lastSyncAt;
    private String error;
}

// CreateClickUpTaskRequest.java
@Data
public class CreateClickUpTaskRequest {
    @NotBlank
    private String name;
    
    private String description;
    private UUID customerId;
    private UUID interactionId;
    private String listId;
    private LocalDate dueDate;
    private Integer priority; // 1=urgent, 2=high, 3=normal, 4=low
}

// ClickUpTaskResponse.java
@Data
@Builder
public class ClickUpTaskResponse {
    private String taskId;
    private String name;
    private String description;
    private String url;
    private String status;
    private Integer priority;
    private UUID linkedCustomerId;
    private UUID linkedInteractionId;
    private LocalDate dueDate;
    private LocalDateTime createdAt;
}

// ZegoTokenResponse.java
@Data
@Builder
public class ZegoTokenResponse {
    private String token;
    private long appID;
    private String serverWSS;
    private String serverWSSBackup;
    private LocalDateTime expiresAt;
}

// ConfigureTelegramRequest.java
@Data
public class ConfigureTelegramRequest {
    @NotBlank
    private String botToken;
    
    private String welcomeMessage;
    private boolean autoReplyEnabled = true;
}

// TelegramConfigResponse.java
@Data
@Builder
public class TelegramConfigResponse {
    private String botUsername;
    private String webhookUrl;
    private String welcomeMessage;
    private boolean autoReplyEnabled;
}
*/

File: backend/src/main/java/com/neobit/crm/controller/IntegrationController.kt
package com.neobit.crm.controller

import com.neobit.crm.dto.ApiResponse
import com.neobit.crm.integration.gmail.*
import com.neobit.crm.integration.calendar.*
import com.neobit.crm.integration.telegram.*
import com.neobit.crm.integration.jira.*
import com.neobit.crm.integration.telephony.*
import com.neobit.crm.security.TenantContext
import com.neobit.crm.security.UserPrincipal
import com.neobit.crm.service.*
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.*
import java.time.ZonedDateTime
import java.util.*

@RestController
@RequestMapping("/api/integrations")
class IntegrationController(
    private val gmailService: GmailService,
    private val calendarService: CalendarService,
    private val telegramService: TelegramService,
    private val jiraService: JiraService,
    private val telephonyService: TelephonyService,
    private val integrationConfigService: IntegrationConfigService
) {
    
    // ============ Integration Config Management ============
    
    @GetMapping("/status")
    fun getAllIntegrationStatus(): ResponseEntity<ApiResponse<List<IntegrationStatus>>> {
        val statuses = integrationConfigService.getAllIntegrations()
        return ResponseEntity.ok(ApiResponse.success(statuses))
    }
    
    @GetMapping("/status/{integrationType}")
    fun getIntegrationStatus(
        @PathVariable integrationType: String
    ): ResponseEntity<ApiResponse<IntegrationStatus>> {
        val status = integrationConfigService.getIntegration(integrationType)
        return if (status != null) {
            ResponseEntity.ok(ApiResponse.success(status))
        } else {
            ResponseEntity.ok(ApiResponse.success(IntegrationStatus(
                integrationType = integrationType,
                isEnabled = false,
                isConfigured = false,
                lastSyncAt = null,
                syncStatus = null
            )))
        }
    }
    
    @PostMapping("/config/jira")
    fun saveJiraConfig(
        @RequestBody request: JiraConfigRequest
    ): ResponseEntity<ApiResponse<IntegrationStatus>> {
        val status = integrationConfigService.saveJiraConfig(request)
        return ResponseEntity.ok(ApiResponse.success(status, "Jira configuration saved"))
    }
    
    @PostMapping("/config")
    fun saveIntegrationConfig(
        @RequestBody request: SaveIntegrationConfigRequest
    ): ResponseEntity<ApiResponse<IntegrationStatus>> {
        val status = integrationConfigService.saveIntegrationConfig(request)
        return ResponseEntity.ok(ApiResponse.success(status, "Integration configuration saved"))
    }
    
    @PostMapping("/{integrationType}/enable")
    fun enableIntegration(
        @PathVariable integrationType: String
    ): ResponseEntity<ApiResponse<IntegrationStatus>> {
        val status = integrationConfigService.enableIntegration(integrationType)
        return if (status != null) {
            ResponseEntity.ok(ApiResponse.success(status, "Integration enabled"))
        } else {
            ResponseEntity.badRequest().body(ApiResponse.error("Integration not configured"))
        }
    }
    
    @PostMapping("/{integrationType}/disable")
    fun disableIntegration(
        @PathVariable integrationType: String
    ): ResponseEntity<ApiResponse<IntegrationStatus>> {
        val status = integrationConfigService.disableIntegration(integrationType)
        return if (status != null) {
            ResponseEntity.ok(ApiResponse.success(status, "Integration disabled"))
        } else {
            ResponseEntity.badRequest().body(ApiResponse.error("Integration not found"))
        }
    }
    
    @DeleteMapping("/{integrationType}")
    fun deleteIntegration(
        @PathVariable integrationType: String
    ): ResponseEntity<ApiResponse<Unit>> {
        val deleted = integrationConfigService.deleteIntegration(integrationType)
        return if (deleted) {
            ResponseEntity.ok(ApiResponse.success(Unit, "Integration deleted"))
        } else {
            ResponseEntity.badRequest().body(ApiResponse.error("Integration not found"))
        }
    }
    
    @PostMapping("/jira/test")
    fun testJiraConnection(): ResponseEntity<ApiResponse<Map<String, Any>>> {
        return try {
            val tenantId = TenantContext.getCurrentTenant()
            val config = jiraService.getJiraConfig(tenantId)
            if (config != null) {
                // Try to fetch a single issue to test connection
                val issues = jiraService.searchIssues(tenantId, "project IS NOT EMPTY", 1)
                ResponseEntity.ok(ApiResponse.success(mapOf(
                    "connected" to true,
                    "baseUrl" to config.baseUrl,
                    "issueCount" to issues.size
                ), "Jira connection successful"))
            } else {
                ResponseEntity.ok(ApiResponse.success(mapOf(
                    "connected" to false,
                    "message" to "Jira not configured"
                )))
            }
        } catch (e: Exception) {
            ResponseEntity.ok(ApiResponse.success(mapOf(
                "connected" to false,
                "error" to (e.message ?: "Connection failed")
            )))
        }
    }
    
    @GetMapping("/jira/projects")
    fun getJiraProjects(): ResponseEntity<ApiResponse<List<Map<String, String>>>> {
        val tenantId = TenantContext.getCurrentTenant()
        val config = jiraService.getJiraConfig(tenantId)
        
        if (config == null) {
            return ResponseEntity.badRequest().body(ApiResponse.error("Jira not configured"))
        }
        
        // Get projects from Jira - simplified, would need to implement in JiraService
        // For now return empty list, the actual project list would be fetched
        return ResponseEntity.ok(ApiResponse.success(emptyList()))
    }
    
    // ============ Gmail Endpoints ============
    
    @GetMapping("/gmail/messages")
    fun listEmails(
        @AuthenticationPrincipal user: UserPrincipal,
        @RequestParam(defaultValue = "20") maxResults: Int,
        @RequestParam(required = false) query: String?
    ): ResponseEntity<ApiResponse<List<EmailMessage>>> {
        val emails = gmailService.listMessages(
            TenantContext.getCurrentTenant(),
            user.id,
            maxResults,
            query
        )
        return ResponseEntity.ok(ApiResponse.success(emails))
    }
    
    @PostMapping("/gmail/send")
    fun sendEmail(
        @AuthenticationPrincipal user: UserPrincipal,
        @RequestBody request: SendEmailRequest
    ): ResponseEntity<ApiResponse<Map<String, String>>> {
        val messageId = gmailService.sendEmail(
            TenantContext.getCurrentTenant(),
            user.id,
            request
        )
        return if (messageId != null) {
            ResponseEntity.ok(ApiResponse.success(mapOf("messageId" to messageId)))
        } else {
            ResponseEntity.badRequest().body(ApiResponse.error("Failed to send email"))
        }
    }
    
    @PostMapping("/gmail/sync")
    fun syncEmails(
        @AuthenticationPrincipal user: UserPrincipal,
        @RequestParam(defaultValue = "newer_than:7d") query: String
    ): ResponseEntity<ApiResponse<List<EmailMessage>>> {
        val emails = gmailService.syncEmails(
            TenantContext.getCurrentTenant(),
            user.id,
            query
        )
        return ResponseEntity.ok(ApiResponse.success(emails, "Synced ${emails.size} emails"))
    }
    
    // ============ Calendar Endpoints ============
    
    @GetMapping("/calendar/events")
    fun listCalendarEvents(
        @AuthenticationPrincipal user: UserPrincipal,
        @RequestParam(required = false) maxResults: Int?
    ): ResponseEntity<ApiResponse<List<CalendarEvent>>> {
        val events = calendarService.listEvents(
            TenantContext.getCurrentTenant(),
            user.id,
            maxResults = maxResults ?: 50
        )
        return ResponseEntity.ok(ApiResponse.success(events))
    }
    
    @GetMapping("/calendar/events/{eventId}")
    fun getCalendarEvent(
        @AuthenticationPrincipal user: UserPrincipal,
        @PathVariable eventId: String
    ): ResponseEntity<ApiResponse<CalendarEvent>> {
        val event = calendarService.getEvent(
            TenantContext.getCurrentTenant(),
            user.id,
            eventId
        )
        return if (event != null) {
            ResponseEntity.ok(ApiResponse.success(event))
        } else {
            ResponseEntity.notFound().build()
        }
    }
    
    @PostMapping("/calendar/events")
    fun createCalendarEvent(
        @AuthenticationPrincipal user: UserPrincipal,
        @RequestBody request: CreateEventRequest
    ): ResponseEntity<ApiResponse<CalendarEvent>> {
        val event = calendarService.createEvent(
            TenantContext.getCurrentTenant(),
            user.id,
            request
        )
        return if (event != null) {
            ResponseEntity.ok(ApiResponse.success(event, "Event created"))
        } else {
            ResponseEntity.badRequest().body(ApiResponse.error("Failed to create event"))
        }
    }
    
    @PutMapping("/calendar/events/{eventId}")
    fun updateCalendarEvent(
        @AuthenticationPrincipal user: UserPrincipal,
        @PathVariable eventId: String,
        @RequestBody request: CreateEventRequest
    ): ResponseEntity<ApiResponse<CalendarEvent>> {
        val event = calendarService.updateEvent(
            TenantContext.getCurrentTenant(),
            user.id,
            eventId,
            request
        )
        return if (event != null) {
            ResponseEntity.ok(ApiResponse.success(event, "Event updated"))
        } else {
            ResponseEntity.badRequest().body(ApiResponse.error("Failed to update event"))
        }
    }
    
    @DeleteMapping("/calendar/events/{eventId}")
    fun deleteCalendarEvent(
        @AuthenticationPrincipal user: UserPrincipal,
        @PathVariable eventId: String
    ): ResponseEntity<ApiResponse<Unit>> {
        val deleted = calendarService.deleteEvent(
            TenantContext.getCurrentTenant(),
            user.id,
            eventId
        )
        return if (deleted) {
            ResponseEntity.ok(ApiResponse.success(Unit, "Event deleted"))
        } else {
            ResponseEntity.badRequest().body(ApiResponse.error("Failed to delete event"))
        }
    }
    
    @PostMapping("/calendar/sync")
    fun syncCalendar(
        @AuthenticationPrincipal user: UserPrincipal
    ): ResponseEntity<ApiResponse<List<CalendarEvent>>> {
        val events = calendarService.syncEvents(
            TenantContext.getCurrentTenant(),
            user.id
        )
        return ResponseEntity.ok(ApiResponse.success(events, "Synced ${events.size} events"))
    }
    
    // ============ Telegram Endpoints ============
    
    @PostMapping("/telegram/send")
    fun sendTelegramMessage(
        @RequestBody request: SendMessageRequest
    ): ResponseEntity<ApiResponse<Unit>> {
        val sent = telegramService.sendMessage(TenantContext.getCurrentTenant(), request)
        return if (sent) {
            ResponseEntity.ok(ApiResponse.success(Unit, "Message sent"))
        } else {
            ResponseEntity.badRequest().body(ApiResponse.error("Failed to send message"))
        }
    }
    
    @GetMapping("/telegram/updates")
    fun getTelegramUpdates(
        @RequestParam(required = false) offset: Long?,
        @RequestParam(defaultValue = "100") limit: Int
    ): ResponseEntity<ApiResponse<List<TelegramUpdate>>> {
        val updates = telegramService.getUpdates(TenantContext.getCurrentTenant(), offset, limit)
        return ResponseEntity.ok(ApiResponse.success(updates))
    }
    
    @PostMapping("/telegram/webhook")
    fun setTelegramWebhook(
        @RequestBody body: Map<String, String>
    ): ResponseEntity<ApiResponse<Unit>> {
        val webhookUrl = body["webhookUrl"] ?: return ResponseEntity.badRequest()
            .body(ApiResponse.error("webhookUrl is required"))
        
        val success = telegramService.setWebhook(TenantContext.getCurrentTenant(), webhookUrl)
        return if (success) {
            ResponseEntity.ok(ApiResponse.success(Unit, "Webhook set"))
        } else {
            ResponseEntity.badRequest().body(ApiResponse.error("Failed to set webhook"))
        }
    }
    
    @DeleteMapping("/telegram/webhook")
    fun deleteTelegramWebhook(): ResponseEntity<ApiResponse<Unit>> {
        val success = telegramService.deleteWebhook(TenantContext.getCurrentTenant())
        return if (success) {
            ResponseEntity.ok(ApiResponse.success(Unit, "Webhook deleted"))
        } else {
            ResponseEntity.badRequest().body(ApiResponse.error("Failed to delete webhook"))
        }
    }
    
    @GetMapping("/telegram/bot-info")
    fun getTelegramBotInfo(): ResponseEntity<ApiResponse<Map<String, Any>>> {
        val botInfo = telegramService.getBotInfo(TenantContext.getCurrentTenant())
        return if (botInfo != null) {
            ResponseEntity.ok(ApiResponse.success(botInfo))
        } else {
            ResponseEntity.badRequest().body(ApiResponse.error("Failed to get bot info or bot not configured"))
        }
    }
    
    @PostMapping("/telegram/chat-ids")
    fun configureTelegramChatIds(
        @RequestBody body: Map<String, Any>
    ): ResponseEntity<ApiResponse<IntegrationStatus>> {
        val chatIds = body["chatIds"] as? List<*>
        if (chatIds == null) {
            return ResponseEntity.badRequest().body(ApiResponse.error("chatIds is required"))
        }
        
        val tenantId = TenantContext.getCurrentTenant()
        val status = integrationConfigService.saveTelegramChatIds(tenantId, chatIds.mapNotNull { 
            when (it) {
                is Number -> it.toLong()
                is String -> it.toLongOrNull()
                else -> null
            }
        })
        
        return ResponseEntity.ok(ApiResponse.success(status, "Telegram chat IDs configured"))
    }
    
    @PostMapping("/telegram/test")
    fun sendTestTelegramMessage(
        @RequestBody body: Map<String, Any>
    ): ResponseEntity<ApiResponse<Unit>> {
        val chatId = (body["chatId"] as? Number)?.toLong()
        val message = body["message"] as? String
        
        if (chatId == null || message == null) {
            return ResponseEntity.badRequest().body(ApiResponse.error("chatId and message are required"))
        }
        
        val tenantId = TenantContext.getCurrentTenant()
        val botToken = telegramService.getBotToken(tenantId)
        
        if (botToken == null) {
            return ResponseEntity.badRequest().body(ApiResponse.error("Telegram bot not configured"))
        }
        
        val sent = telegramService.sendMessageWithToken(botToken, chatId, message)
        return if (sent) {
            ResponseEntity.ok(ApiResponse.success(Unit, "Test message sent"))
        } else {
            ResponseEntity.badRequest().body(ApiResponse.error("Failed to send test message"))
        }
    }
    
    // ============ Jira Endpoints ============
    
    @PostMapping("/jira/issues")
    fun createJiraIssue(
        @RequestBody request: CreateJiraIssueRequest
    ): ResponseEntity<ApiResponse<JiraIssue>> {
        val issue = jiraService.createIssue(TenantContext.getCurrentTenant(), request)
        return if (issue != null) {
            ResponseEntity.ok(ApiResponse.success(issue, "Issue created"))
        } else {
            ResponseEntity.badRequest().body(ApiResponse.error("Failed to create issue"))
        }
    }
    
    @GetMapping("/jira/issues/{issueKey}")
    fun getJiraIssue(
        @PathVariable issueKey: String
    ): ResponseEntity<ApiResponse<JiraIssue>> {
        val issue = jiraService.getIssue(TenantContext.getCurrentTenant(), issueKey)
        return if (issue != null) {
            ResponseEntity.ok(ApiResponse.success(issue))
        } else {
            ResponseEntity.notFound().build()
        }
    }
    
    @GetMapping("/jira/search")
    fun searchJiraIssues(
        @RequestParam jql: String,
        @RequestParam(defaultValue = "50") maxResults: Int
    ): ResponseEntity<ApiResponse<List<JiraIssue>>> {
        val issues = jiraService.searchIssues(TenantContext.getCurrentTenant(), jql, maxResults)
        return ResponseEntity.ok(ApiResponse.success(issues))
    }
    
    @PutMapping("/jira/issues/{issueKey}")
    fun updateJiraIssue(
        @PathVariable issueKey: String,
        @RequestBody request: UpdateJiraIssueRequest
    ): ResponseEntity<ApiResponse<JiraIssue>> {
        val issue = jiraService.updateIssue(TenantContext.getCurrentTenant(), issueKey, request)
        return if (issue != null) {
            ResponseEntity.ok(ApiResponse.success(issue, "Issue updated"))
        } else {
            ResponseEntity.badRequest().body(ApiResponse.error("Failed to update issue"))
        }
    }
    
    @PostMapping("/jira/issues/{issueKey}/transition")
    fun transitionJiraIssue(
        @PathVariable issueKey: String,
        @RequestBody body: Map<String, String>
    ): ResponseEntity<ApiResponse<Unit>> {
        val status = body["status"] ?: return ResponseEntity.badRequest()
            .body(ApiResponse.error("status is required"))
        
        val success = jiraService.transitionIssue(TenantContext.getCurrentTenant(), issueKey, status)
        return if (success) {
            ResponseEntity.ok(ApiResponse.success(Unit, "Issue transitioned"))
        } else {
            ResponseEntity.badRequest().body(ApiResponse.error("Failed to transition issue"))
        }
    }
    
    @PostMapping("/jira/issues/{issueKey}/comment")
    fun addJiraComment(
        @PathVariable issueKey: String,
        @RequestBody body: Map<String, String>
    ): ResponseEntity<ApiResponse<Unit>> {
        val comment = body["comment"] ?: return ResponseEntity.badRequest()
            .body(ApiResponse.error("comment is required"))
        
        val success = jiraService.addComment(TenantContext.getCurrentTenant(), issueKey, comment)
        return if (success) {
            ResponseEntity.ok(ApiResponse.success(Unit, "Comment added"))
        } else {
            ResponseEntity.badRequest().body(ApiResponse.error("Failed to add comment"))
        }
    }
    
    // ============ Telephony Endpoints ============
    
    @PostMapping("/telephony/call")
    fun initiateCall(
        @AuthenticationPrincipal user: UserPrincipal,
        @RequestBody request: CallRequest
    ): ResponseEntity<ApiResponse<CallResult>> {
        val result = telephonyService.initiateCall(
            TenantContext.getCurrentTenant(),
            user.id,
            request
        )
        return if (result != null) {
            ResponseEntity.ok(ApiResponse.success(result, "Call initiated"))
        } else {
            ResponseEntity.badRequest().body(ApiResponse.error("Failed to initiate call"))
        }
    }
    
    @GetMapping("/telephony/calls/{callId}")
    fun getCallStatus(
        @PathVariable callId: String
    ): ResponseEntity<ApiResponse<CallResult>> {
        val result = telephonyService.getCallStatus(TenantContext.getCurrentTenant(), callId)
        return if (result != null) {
            ResponseEntity.ok(ApiResponse.success(result))
        } else {
            ResponseEntity.notFound().build()
        }
    }
    
    @GetMapping("/telephony/calls")
    fun listRecentCalls(
        @RequestParam(defaultValue = "50") limit: Int
    ): ResponseEntity<ApiResponse<List<CallResult>>> {
        val calls = telephonyService.listRecentCalls(TenantContext.getCurrentTenant(), limit)
        return ResponseEntity.ok(ApiResponse.success(calls))
    }
}

